### PR TITLE
Don't allocate in NullStream.CopyToAsync

### PR DESCRIPTION
**Pull request by jamesqo**
This is basically a port of the changes in dotnet/corefx#7782 to CoreCLR. Here are the changes:

- A `TaskCache` class has been introduced to hold cached tasks for the values 0, null, and the empty string.

- A few classes have been optimized in their `Read` methods to return the task holding 0 in `TaskCache`, instead of allocating a new `Task` every time.

- Methods on the null-object readers/writers (e.g. `Stream.Null`, `StreamReader.Null`) have been overridden to more efficiently noop/return a completed Task, since reading/writing from them isn't supposed to do anything.

cc @jkotas @justinvp

### CODE REVIEW

**Comment by jkotas** [comment]
@jamesqo Have you seen these as a bottleneck in any real program?

It is not clear to me that these optimizations are worth the extra code. Extra code or caches are not free - they take memory too, and moreover this memory stays allocated till the program exits.

E.g. Reading stream asynchronously typically requires N read calls, and it is generally not a super cheap operation. Saving one allocation in the whole process does not look like something that would move the needle anywhere.

@stephentoub @AlexGhiondea

File: src/mscorlib/src/System/IO/Stream.cs [hunk 1]
```diff
@@ -900,6 +900,13 @@ private sealed class NullStream : Stream
                 get { return 0; }
                 set {}
             }
+            
+            public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                if (cancellationToken.IsCancellationRequested)
```
Comment (jkotas) [inline src/mscorlib/src/System/IO/Stream.cs]: I cannot imagine a case where this optimized implementation would make a difference.

File: src/mscorlib/src/System/IO/TextWriter.cs [hunk 2]
```diff
@@ -606,30 +606,69 @@ public virtual Task FlushAsync()
...
+            public override void Write(decimal value) { }
+            public override void Write(double value) { }
+            public override void Write(float value) { }
+            public override void Write(int value) { }
+            public override void Write(long value) { }
+            public override void Write(object value) { }
+            public override void Write(string format, object arg0) { }
+            public override void Write(string format, object arg0, object arg1) { }
+            public override void Write(string format, object arg0, object arg1, object arg2) { }
+            public override void Write(string format, params object[] arg) { }
```
Comment (jkotas) [inline src/mscorlib/src/System/IO/TextWriter.cs]: This has a subtle compat issue: There is no argument validation.

File: src/mscorlib/src/System/IO/BufferedStream.cs [hunk 3]
```diff
@@ -641,7 +641,9 @@ public BufferedStream(Stream stream)
         if (t != null && t.Result == val)
             return t;
         
-        t = Task.FromResult<Int32>(val);
+        t = val == 0 ?
+            TaskCache.Zero : // Return a cached Task if EOF has been reached
+            Task.FromResult<int>(val);
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/BufferedStream.cs]: This, and the other places in this PR like it, is adding an extra branch to every read, and the only time it'll help avoid an allocation is on the very last read to the stream.  Typically streams are read from many times, especially in the case of a buffered stream, so this is adding such branches across all read calls in exchange for saving a single allocation for the lifetime of the stream.

File: src/mscorlib/src/System/IO/TextWriter.cs [hunk 4]
```diff
@@ -606,30 +606,69 @@ public virtual Task FlushAsync()
...
+            {
             }
 
-            [SuppressMessage("Microsoft.Contracts", "CC1055")]  // Skip extra error checking to avoid *potential* AppCompat problems.
-            public override void Write(char[] buffer, int index, int count) {
-            }
-            
-            public override void Write(String value) {
-            }
+            public override Encoding Encoding => Encoding.Unicode;
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/TextWriter.cs]: This is a breaking change.

File: src/mscorlib/src/System/Threading/Tasks/TaskCache.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Threading.Tasks
+{
+    internal static class TaskCache
```
Comment (stephentoub) [inline src/mscorlib/src/System/Threading/Tasks/TaskCache.cs]: There's already a task cache in mscorlib, e.g.
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs#L801
We shouldn't be adding a separate one.

File: src/mscorlib/src/System/IO/Stream.cs (hunk 1 above)
Comment (jamesqo) [inline src/mscorlib/src/System/IO/Stream.cs]: @jkotas Well maybe, but `Stream.Null` already overrides `ReadAsync`/`WriteAsync` for perf, so why not `CopyToAsync` as well?

File: src/mscorlib/src/System/IO/TextWriter.cs (hunk 4 above)
Comment (jamesqo) [inline src/mscorlib/src/System/IO/TextWriter.cs]: Slipped past me that the original implementation was returning `Encoding.Default`; my bad.

File: src/mscorlib/src/System/IO/Stream.cs (hunk 1 above)
Comment (stephentoub) [inline src/mscorlib/src/System/IO/Stream.cs]: > but Stream.Null already overrides ReadAsync/WriteAsync for perf, so why not CopyToAsync as well?

If it didn't override ReadAsync/WriteAsync, the base implementations would end up allocating.  That's not the case with CopyToAsync, where the base implementation will call ReadAsync and return immediately, without allocating or incurring much cost at all.

File: src/mscorlib/src/System/IO/TextWriter.cs (hunk 2 above)
Comment (jamesqo) [inline src/mscorlib/src/System/IO/TextWriter.cs]: @jkotas OK, will revert the changes in this file then (as they take up a lot of code as well).

**Comment by jamesqo** [comment]
@stephentoub @jkotas Since the `CopyToAsync` changes seem to provide the only significant perf benefit in this PR (as @stephentoub pointed out, by default it will [allocate](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/IO/Stream.cs#L164) a giant 81920-sized byte array), I've decided to reduce my PR to just those changes. I've also added argument validation for compat; could you please re-PTAL?

**Comment by jamesqo** [comment]
@dotnet-bot test this please

**Comment by jkotas** [comment]
Would it be better to just call the base class with buffer size of 1. Like `CopyToAsync(destination, Math.Min(bufferSize, 1), cancellationToken)` ? There would be still some tiny allocation left, but I think it is fine cost to pay to avoid the duplicated error handling.

BTW: I have noticed that there are `CopyToAsyncInternal` and `InternalCopyTo` methods for no good reason. It would be nice to fold them into the public entrypoints and remove the duplicated error handling in `CopyTo(Stream destination)` - if you would like to do it in this PR, it would be fine.

**Comment by stephentoub** [comment]
> BTW: I have noticed that there are CopyToAsyncInternal and InternalCopyTo methods for no good reason

For CopyToAsyncInternal there is good reason: as with iterators, exceptions in async methods do not propagate synchronously out of the call.

**Comment by stephentoub** [comment]
> but I think it is fine cost to pay to avoid the duplicated error handling.

Another option would be to move all of that error handling verification into a private method on Stream, which NullStream would also be able to use as a private nested type.

But, I'd also be fine with the Math.Min option Jan proposes.

Thanks, @jamesqo.

**Comment by jamesqo** [comment]
@stephentoub and Jan: Thanks for the feedback, I've addressed all of the points you guys have made.

> But, yeah, for InternalCopyto, seems unnecessary.

Just pushed a commit that removes this helper method, but for some reason I was getting a strange error with `BclRewriter` when I tried to build locally beforehand. It says this:

```
  mscorlib -> C:\Users\James\Code\git\coreclr\bin\obj\Windows_NT.x64.Release\mscorlib.dll
BclRewriter : error BR0000: cannot load member Method : InternalCopyTo(System.IO.Stream,System.Int32) from class System.IO.Stream.  No members of that name were found in the assembly. [C:\Users\James\Code\git\coreclr\src\mscorlib\mscorlib.csproj]
```

Not sure what exactly is causing this; let's see if the CI passes.

**Comment by justinvp** [comment]
> Just pushed a commit that removes this helper method, but for some reason I was getting a strange error with BclRewriter when I tried to build locally beforehand

You need to delete the method from model.xml: https://github.com/dotnet/coreclr/blob/b82077e95ca8834ee97f763c1382d273d5fc6767/src/mscorlib/model.xml#L8060

I'm not sure if you need to add the new method or if it'll be pulled in automatically as part of the dependency closure.

File: src/mscorlib/src/System/IO/Stream.cs [hunk 6]
```diff
@@ -871,6 +821,23 @@ internal static void BlockingEndWrite(IAsyncResult asyncResult)
         {
             SynchronousAsyncResult.EndWrite(asyncResult);
         }
+        
+        internal void ValidateCopyToArguments(Stream destination, int bufferSize)
```
Comment (justinvp) [inline src/mscorlib/src/System/IO/Stream.cs]: Could this be `private`?

Comment (justinvp) [inline src/mscorlib/src/System/IO/Stream.cs]: I'm guessing we don't really care about code contracts these days, but should this method have `[ContractArgumentValidator]`?

File: src/mscorlib/src/System/IO/Stream.cs [hunk 7]
```diff
@@ -900,6 +867,17 @@ private sealed class NullStream : Stream
...
+            
+            public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                // Validate arguments here for compat, since previously this method
+                // was inherited from Stream (which did check its arguments).
+                ValidateCopyToArguments(destination, bufferSize);
+                
+                if (cancellationToken.IsCancellationRequested)
+                    return Task.FromCancellation(cancellationToken);
+                return Task.CompletedTask;
```
Comment (justinvp) [inline src/mscorlib/src/System/IO/Stream.cs]: This could be simplified:

```csharp
return cancellationToken.IsCancellationRequested ? Task.FromCancellation(cancellationToken) : Task.CompletedTask;
```

File: src/mscorlib/src/System/IO/Stream.cs [hunk 8]
```diff
@@ -172,57 +160,19 @@ public virtual Task CopyToAsync(Stream destination, Int32 bufferSize, Cancellati
...
-                throw new ObjectDisposedException("destination", Environment.GetResourceString("ObjectDisposed_StreamClosed"));
-            if (!CanRead)
-                throw new NotSupportedException(Environment.GetResourceString("NotSupported_UnreadableStream"));
-            if (!destination.CanWrite)
-                throw new NotSupportedException(Environment.GetResourceString("NotSupported_UnwritableStream"));
-            Contract.EndContractBlock();
-
-            InternalCopyTo(destination, _DefaultCopyBufferSize);
-        }
+        public void CopyTo(Stream destination) =>
```
Comment (jkotas) [inline src/mscorlib/src/System/IO/Stream.cs]: It would look better to do it old fashioned way to maintain consistent style within the file.

File: src/mscorlib/src/System/IO/Stream.cs (hunk 6 above)
Comment (jamesqo) [inline src/mscorlib/src/System/IO/Stream.cs]: @justinvp Doing a GitHub search for "ContractArgumentValidator" in this repo doesn't turn up anything, so I'm guessing no as well. @jkotas let me know otherwise.

Comment (jamesqo) [inline src/mscorlib/src/System/IO/Stream.cs]: Personally, I think it would be best to leave this `internal` since there are other subclasses of `Stream` that could benefit from this in a future PR, e.g. `MemoryStream` [here](https://github.com/dotnet/coreclr/blob/ebb839f3218a9410f76f4595574155304c5e101e/src/mscorlib/src/System/IO/MemoryStream.cs#L406).

**Comment by jamesqo** [comment]
@justinvp Does every method/type in mscorlib, public or not, have to be defined in `model.xml`?

**Comment by justinvp** [comment]
> Does every method/type in mscorlib, public or not, have to be defined in model.xml?

My understanding is that public/protected APIs intended to be included in the subsetted mscorlib for CoreCLR need to be defined in model.xml. I'm not sure when it's necessary to specify internals in model.xml. For example, I'm not sure why `InternalCopyTo` was in model.xml but `CopyToAsyncInternal` is not. I was under the impression that the rewriter would do a dependency analysis and keep internal dependencies automatically, but maybe the tool needs manual help sometimes? @jkotas or @stephentoub, do you know?

**Comment by jkotas** [comment]
> My understanding is that public/protected APIs intended to be included in the subsetted mscorlib for CoreCLR need to be defined in model.xml.

Correct.

> I'm not sure when it's necessary to specify internals in model.xml. 

It is only necessary for internals that something depends on (e.g. internals referenced from vm).

> For example, I'm not sure why  InternalCopyTo  was in model.xml 

I am not able to find why it was added originally (i looked back till 2010), it might have been just a mistake. I do see any reason why it should be still needed - it should be safe to remove it.

> I was under the impression that the rewriter would do a dependency analysis and keep internal dependencies automatically

Correct.

File: src/mscorlib/src/System/IO/Stream.cs (hunk 6 above)
Comment (jkotas) [inline src/mscorlib/src/System/IO/Stream.cs]: > Doing a GitHub search for "ContractArgumentValidator" in this repo doesn't turn up anything

Correct. We are not maintaining the code contracts annotation anymore.

**Comment by jamesqo** [comment]
@jkotas PR feedback has been addressed.

**Comment by jamesqo** [comment]
test osx x64 checked build and test

**Comment by jamesqo** [comment]
Test OSX x64 Checked Build and Test

**Comment by jkotas** [comment]
LGTM - it should be ready to be merged.

The OSX failure is infrastructure problem.

**Comment by stephentoub** [comment]
LGTM. Thanks, @jamesqo.

**Comment by jamesqo** [comment]
@jkotas @stephentoub Awesome, I've squashed the commit.

BTW, to make your guys' lives easier next time, I believe GitHub has recently added a way to force-squash pull requests via the web interface: https://github.com/blog/2141-squash-your-commits Haven't tried it out for myself yet, though.

**Comment by jkotas** [comment]
Thank you!

> a way to force-squash pull requests via the web interface

We need to make fixes to parts of the engineering system like the TFS mirroring before we can start using it.

**Comment by jkotas** [comment]
@jamesqo Would you mind porting this change to the CoreRT version of Stream.cs in corefx repo? Also, it would be nice to add test coverage for the overridden CopyToAsync on NullStream to corefx repo as well.

**Comment by jamesqo** [comment]
@jkotas Sure, no problem. I'll open a pull request sometime today/tomorrow when I have time. :smile:
